src/Polysemy/History.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+module Polysemy.History
+  ( -- * Effect
+    History (..)
+
+    -- * Actions
+  , undo
+  , redo
+  , putAndForget
+
+    -- * Interpretations
+  , runHistory
+  ) where
+
+import Polysemy
+import Polysemy.State
+import Polysemy.Tagged
+
+
+data History s m a where
+  Undo :: History s m ()
+  Redo :: History s m ()
+  PutAndForget :: s -> History s m ()
+
+makeSem ''History
+
+
+runHistory
+    :: forall s r a
+     . Member (State s) r
+    => Sem (History s ': r) a
+    -> Sem r a
+runHistory m = do
```

Comment:
Is `historyToIO` possible? Feels like it should be.

Reply:
with what semantics?

Reply:
Global-to-everything semantics; i.e. `State (Zipper s)` is interpreted via `stateToIO`. It's the only way we get consistency if the targeted `State s` is interpreted via `stateToIO` (since `get`/`put` of `State (Zipper s)` can get reverted by IO exceptions, but not of the `State s`).

Then again, `historyToIO` then becomes inconsistent if `State s` is interpreted via `runState`. Maybe it'd be too messy.

Reply:
Maybe there's a more general solution here which is to be able to steal an interpretation?

Reply:
I don't think it's possible to interpret `History` solely using `State s`.

However, I've previously entertained the notion of a "StateManager" effect, like this:

```haskell
data StateManager k m a where
  StateCarrier :: k a -> StateManager k m a
  ManageState :: Member (StateManager k) r => s -> Sem (State s ': r) a -> StateManager k m (s, a)

makeSem ''StateManager

runStateManager :: Sem (StateManager (Const Void) ': r) a -> Sem r a
runStateManager = interpretH $ \case
  StateCarrier (Cont m) -> absurd m
  ManageState s m -> runT (runState s m) >>= raise . runStateManager

stateManagerToIO :: Member (Embed IO) r => Sem (StateManager IO ': r) a -> Sem r a
stateManagerToIO = interpretH $ \case
  StateCarrier m -> embed m
  ManageState s m -> do
    ref <- embed $ newIORef s
    m'<- runT $ ($ m) $ interpret $ \case
      Put s' -> stateCarrier $ writeIORef ref s'
      Get -> stateCarrier $ readIORef ref
    res <- raise $ stateManagerToIO m'
    end <- embed $ readIORef ref
    return $ fmap (end,) res
```

It's just that I've been looking for a formulation of it that isn't, y'know, disgusting.